ush/python/pygw/src/pygw/schema.py

```diff
@@ -0,0 +1,891 @@
+
+
+class Forbidden(Hook):
+    def __init__(self, *args, **kwargs):
+        kwargs["handler"] = self._default_function
+        super(Forbidden, self).__init__(*args, **kwargs)
+
+    @staticmethod
+    def _default_function(nkey, data, error):
+        raise SchemaForbiddenKeyError(
+            "Forbidden key encountered: %r in %r" % (nkey, data), error)
+
+
+class Literal(object):
+    def __init__(self, value, description=None):
+        self._schema = value
+        self._description = description
+
+    def __str__(self):
+        return self._schema
+
+    def __repr__(self):
+        return 'Literal("' + self.schema + '", description="' + (self.description or "") + '")'
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def schema(self):
+        return self._schema
+
+
+class Const(Schema):
+    def validate(self, data, **kwargs):
+        super(Const, self).validate(data, **kwargs)
+        return data
+
+
+def _callable_str(callable_):
+    if hasattr(callable_, "__name__"):
+        return callable_.__name__
+    return str(callable_)
+
+
+def _plural_s(sized):
+    return "s" if len(sized) > 1 else ""
+
+# ----
+
+# The following functions are specific to global-workflow. The Schema
+# class module was obtained from:
+
+# https://github.com/keleshev/schema/blob/master/schema.py
+
+
+# TODO: Is there a better way to do this?
+logger = Logger("pygw.schema", colored_log=True)
+
+
+@logit(logger)
+def build_schema(yaml_path: str) -> Dict:
+    """
+    Description
+    -----------
+
+    This function parses a YAML-formatted file and defines the
+    respective schema.
+
+    Parameters
+    ----------
+
+    yaml_path: str
+
+        A Python string specifying the path to the YAML-formatted file
+        containing the schema attributes.
+
+    schema_dict: Dict
+
+        A Python dictionary containing the schema.
+
+    """
+
+    # Read the YAML-formatted file containing the schema.
+    data = yaml_file.parse_yaml(path=yaml_path)
+
+    schema_dict = {}
+    for datum in data:
+        data_dict = AttrDict(data[datum])
+
+        # Check whether the variable is optional; proceed accordingly.
+        if "optional" not in data_dict:
+            data_dict.optional = False
+        else:
+            pass
+
+        # Build the schema accordingly.
+        if data_dict.optional:
+            schema_dict[Optional(datum, data_dict.default)
+                        ] = locate(data_dict.type)
+
+        if not data_dict.optional:
+            schema_dict[datum] = locate(data_dict.type)
+
+    return schema_dict
+
+# ----
+
+
+@logit(logger)
+def validate_schema(schema_dict: Dict, cfg: Dict) -> Dict:
+    """
+    # TODO: This method will read the config_obj and the schema defined in `build_schema`; if a value is 
+    not defined in `cfg` and is optional, the default value will be assigned; if a value is not defined 
+    and is not optional, and exception will be raised.
+
+    """
```

Comment:
Not sure this is needed

Reply:
Leaving this for now. We should chat as to what you'd like to see here.

Reply:
ok.